File: src/main/scala/dev/alexnader/framity/block/frame/FrameData.scala
package dev.alexnader.framity.block.frame

import com.mojang.serialization.Dynamic
import dev.alexnader.framity.util.MinecraftUtil.CompoundTagExt
import net.minecraft.block.BlockState
import net.minecraft.nbt.NbtOps
import net.minecraft.item.ItemStack
import net.minecraft.nbt.{CompoundTag, IntTag, ListTag}

import scala.jdk.CollectionConverters._
import scala.language.postfixOps

object FrameData {

  object Sections {

    implicit class SlotSectionConversions(val i: Int) {
      def toRelative(r: Range): Int = i - r.start

      def toAbsolute(r: Range): Int = r.start + i
    }

    val BaseIndex: Int = 0
    val OverlayIndex: Int = 1
    val SpecialIndex: Int = 2

    def fromTag(tag: ListTag): Sections = {
      new Sections(makeSections(tag.iterator.asScala map (_.asInstanceOf[IntTag].getInt)))
    }

    private def makeSections(sizes: Iterator[Int]): Array[Range] = {
      var start = 0
      sizes map { size =>
        val r = start until start + size
        start += size
        r
      } toArray
    }
  }

  class Sections(private val sections: Array[Range]) {
    def this(partCount: Int, otherSizes: Int*) = this(Sections.makeSections(Iterator(partCount) ++ Iterator(partCount) ++ Iterator(SpecialItems.map.size) ++ otherSizes.iterator))

    def apply(sectionIndex: Int): Range = sections(sectionIndex)

    def base: Range = this (Sections.BaseIndex)

    def overlay: Range = this (Sections.OverlayIndex)

    def special: Range = this (Sections.SpecialIndex)

    def itemCount: Int = sections.last.end + 1

    def sectionCount: Int = sections.length

    def itemIndices: Range = 0 until sections.last.end

    def containsSlot(slot: Int): Boolean = 0 <= slot && slot < sections.last.end

    def findSectionIndexOf(absoluteIndex: Int): Int = {
      @scala.annotation.tailrec
      def findSectionIndexOf(i: Int, absoluteIndex: Int): Int = {
        if (absoluteIndex < sections(i).end) {
          i
        } else if (i == sections.length) {
          -1
        } else {
          findSectionIndexOf(i + 1, absoluteIndex)
        }
      }

      findSectionIndexOf(0, absoluteIndex)
    }

    def makeItems: Array[Option[ItemStack]] =
      Array.fill[Option[ItemStack]](itemCount)(None)

    def makeBaseStates: Array[Option[BlockState]] =
      Array.fill[Option[BlockState]](base.size)(None)

    def toTag: ListTag = {
      val tag = new ListTag

      sections foreach (s => tag.add(IntTag.of(s.size)))

      tag
    }

    def canEqual(other: Any): Boolean = other.isInstanceOf[Sections]

    override def equals(other: Any): Boolean = other match {
      case that: Sections =>
        (that canEqual this) &&
          (sections sameElements that.sections)
      case _ => false
    }

    override def hashCode(): Int = {
      val state = Seq(sections)
      state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
    }
  }

  private def itemsFromTag(sections: Sections, tag: ListTag): Array[Option[ItemStack]] = {
    val items = sections.makeItems

    tag.asScala.indices foreach { i =>
      val stackTag = tag.getCompound(i)
      val slot = stackTag.getByte("Slot") & 255
      if (sections containsSlot slot) {
        items(slot) = Some(ItemStack.fromTag(stackTag))
      }
    }

    items
  }

  private def statesFromTag(sections: Sections, tag: ListTag): Array[Option[BlockState]] = {
    val states = sections.makeBaseStates

    tag.asScala.indices foreach { tagIndex =>
      val stateTag = tag.getCompound(tagIndex)
      val realIndex = stateTag.getInt("i")
      states(realIndex) = Some(BlockState.CODEC.decode(new Dynamic(NbtOps.INSTANCE, stateTag)).result.get.getFirst)
    }

    states
  }

  def fromTag(tag: CompoundTag): FrameData = {
    val sections = Sections fromTag tag.getList("format", 3)

    FrameData(sections, itemsFromTag(sections, tag.getList("Items", 10)), statesFromTag(sections, tag.getList("states", 10)))
  }
}

case class FrameData(sections: FrameData.Sections, items: Array[Option[ItemStack]], baseStates: Array[Option[BlockState]]) {
  def this(sections: FrameData.Sections) = this(sections, sections.makeItems, sections.makeBaseStates)

  def toTag: CompoundTag = {
    val tag = new CompoundTag

    tag("format") = sections.toTag

    val itemsTag = items.iterator.zipWithIndex.foldLeft(new ListTag) { case (acc, (stack, slot)) =>
      stack match {
        case Some(stack) =>
          val stackTag = new CompoundTag
          stack.toTag(stackTag)
          stackTag.putByte("Slot", slot.byteValue)
          acc.add(stackTag)
        case None =>
      }
      acc
    }
    if (!itemsTag.isEmpty) {
      tag("Items") = itemsTag
    }

    val statesTag = baseStates.iterator.zipWithIndex.foldLeft(new ListTag) { case (acc, (state, i)) =>
      state match {
        case Some(state) =>
          val stateTag = new CompoundTag
          stateTag.putInt("i", i)
          acc.add(BlockState.CODEC.encode(state, NbtOps.INSTANCE, stateTag).get.left.get)
        case None =>
      }
      acc
    }
    if (!statesTag.isEmpty) {
      tag("states") = statesTag
    }

    tag
  }
}

File: src/main/java/dev/alexnader/framed/block/frame/data/FrameData.java
package dev.alexnader.framed.block.frame.data;

import com.google.common.collect.Streams;
import com.mojang.datafixers.util.Pair;
import com.mojang.serialization.Dynamic;
import net.minecraft.block.BlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.nbt.ListTag;
import net.minecraft.nbt.NbtOps;
import net.minecraft.util.Identifier;

import javax.annotation.Nonnull;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static dev.alexnader.framed.Framed.OVERLAYS;

public class FrameData {
    private static Optional<ItemStack>[] itemsFromTag(final Sections sections, final ListTag tag) {
        final Optional<ItemStack>[] items = sections.makeItems();

        for (int i = 0, size = tag.size(); i < size; i++) {
            final CompoundTag stackTag = tag.getCompound(i);
            final int slot = stackTag.getByte("Slot") & 255;
            if (sections.containsSlot(slot)) {
                items[slot] = Optional.of(ItemStack.fromTag(stackTag));
            }
        }

        return items;
    }

    private static Optional<BlockState>[] baseStatesFromTag(final Sections sections, final ListTag tag) {
        final Optional<BlockState>[] baseStates = sections.makeBaseStates();

        for (int i = 0, size = tag.size(); i < size; i++) {
            final CompoundTag stateTag = tag.getCompound(i);
            final int realIndex = stateTag.getInt("i");
            //noinspection OptionalGetWithoutIsPresent
            baseStates[realIndex] = Optional.of(
                BlockState.CODEC.decode(new Dynamic<>(NbtOps.INSTANCE, stateTag)).result().get().getFirst()
            );
        }

        return baseStates;
    }

    public static FrameData fromTag(final CompoundTag tag) {
        final Sections sections = Sections.fromTag(tag.getList("format", 3));

        return new FrameData(
            sections,
            itemsFromTag(sections, tag.getList("Items", 10)),
            baseStatesFromTag(sections, tag.getList("states", 10))
        );
    }

    private final Sections sections;
    private final Optional<ItemStack>[] items;
    private final Optional<BlockState>[] baseStates;

    public FrameData(final Sections sections, final Optional<ItemStack>[] items, final Optional<BlockState>[] baseStates) {
        this.sections = sections;
        this.items = items;
        this.baseStates = baseStates;
    }

    public FrameData(@Nonnull final Sections sections) {
        this(sections, sections.makeItems(), sections.makeBaseStates());
    }

    public Sections sections() {
        return sections;
    }

    public Optional<ItemStack>[] items() {
        return items;
    }

    public List<Optional<ItemStack>> baseItems() {
        return Arrays.asList(items).subList(sections.base().start(), sections.base().end());
    }

    public List<Optional<ItemStack>> overlayItems() {
        return Arrays.asList(items).subList(sections.overlay().start(), sections.overlay().end());
    }

    public List<Optional<ItemStack>> specialItems() {
        return Arrays.asList(items).subList(sections.special().start(), sections.special().end());
    }

    public Optional<BlockState>[] baseStates() {
        return baseStates;
    }

    public CompoundTag toTag() {
        final CompoundTag tag = new CompoundTag();

        tag.put("format", sections.toTag());

        final ListTag itemsTag = new ListTag();
        for (int i = 0, size = items.length; i < size; i++) {
            final int i2 = i;
            items[i].ifPresent(stack -> {
                final CompoundTag stackTag = new CompoundTag();
                stack.toTag(stackTag);
                stackTag.putByte("Slot", (byte)i2);
                itemsTag.add(stackTag);
            });
        }
        if (!itemsTag.isEmpty()) {
            tag.put("Items", itemsTag);
        }

        final ListTag baseStatesTag = new ListTag();
        for (int i = 0, size = baseStates.length; i < size; i++) {
            final int i2 = i;
            baseStates[i].ifPresent(baseState -> {
                final CompoundTag baseStateTag = new CompoundTag();
                baseStateTag.putInt("i", i2);
                //noinspection OptionalGetWithoutIsPresent
                baseStatesTag.add(
                    BlockState.CODEC.encode(baseState, NbtOps.INSTANCE, baseStateTag).get().left().get()
                );
            });
        }
        if (!baseStatesTag.isEmpty()) {
            tag.put("states", baseStatesTag);
        }

        return tag;
    }

    public List<Pair<Optional<BlockState>, Optional<Identifier>>> toRenderAttachment() {
        //noinspection UnstableApiUsage
        return Streams.zip(
            Arrays.stream(baseStates),
            overlayItems().stream().map(i -> i.flatMap(OVERLAYS::getOverlayId)),
            Pair::new
        ).collect(Collectors.toList());
    }
}
